### ISSUE TITLE
Bug: no-unnecessary-condition false negative inside string and number conditionals

### ISSUE THREAD

**Issue by guilhermesimoes** [typescript-eslint/typescript-eslint#7926]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=CYUwxgNghgTiAEEQBd4CMD2GIC51aSgDsBuAWACgBLAM3gApNsBKeAb0vnloaYlY4UuXMBiIBnbCAB0EDAHNGBeAB8V8AOQ0sG5uSHwAvpWMVKoSLARJURAK4BbPPYdoQMfT3ouBnbnW9HAXgAehD4cQALDDsIYHhQZHBUdxgMGHhI9xA-ETFJJFkFQIdVdS0dPVDwqJi4hJRk%2BFT0zOy-U1NzcGg4RBQI5Bg8cSGqInlPANGYXwMvGeCwiOjY%2BMSmloysuFz4UQkpIsUZss1tDF0SapW69cawFJg07faDTqA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
declare let str: string;
if (str) {
  if (str) {
    console.log(str || 'foo');
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

The error "Unnecessary conditional, value is always truthy" should be reported on lines 3 and 4

### Actual Result

No errors are detected

### Additional Info

I discovered this because I was trying out [eslint-plugin-sonarjs](https://github.com/SonarSource/eslint-plugin-sonarjs), and that plugin has a similar rule called [no-gratuitous-expression](https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/no-gratuitous-expressions.md) that does detect these cases. The code for that rule is available [here](https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/src/rules/no-gratuitous-expressions.ts).

**Comment by bradzacher** [comment]
This is working as expected.
The `no-unnecessary-condition` rule is powered solely by types.

Under-the-hood within TS the `boolean` type is actually defined as a union type (`true | false`) so it follows logically that TS can refine `boolean` to `true` or `false` based on a conditional check.

However other primitives are not and cannot be defined like that. Which means that the result of refining `string` via a conditional is always going to be `string`.
There's talk of "negated types" (eg https://github.com/microsoft/TypeScript/issues/4196) which would allow TS to refine this case to eg `string & not ''` but currently that feature doesn't exist.
TS also doesn't narrow the false branch to values like `''` or `0 | NaN` - presumably for simplicity's sake.

---

The rule is already complicated enough going through all of the logic required to evaluate types - and I don't think there's much value in adding a lot of additional complexity to catch a relatively rare case - especially when there are other rules that already exist and catch the case.
